apps/wwaw/build_wwaw.py

```diff
@@ -0,0 +1,604 @@
+    parsed = urlparse(link)
+    return parsed.scheme in ("http", "https") and base_domain in parsed.netloc
+
+
+def clean_and_extract_text(html):
+    """
+    Cleans and extracts readable text content from HTML.
+
+    Removes non-content elements (e.g., scripts, styles, images), strips unnecessary attributes,
+    extracts visible text from paragraph-level tags, and sanitizes the text by removing URLs,
+    special characters, and non-ASCII content.
+
+    Args:
+        html (str): Raw HTML content of a web page.
+
+    Returns:
+        str: Cleaned and normalized text extracted from the HTML.
+    """
+    soup = BeautifulSoup(html, 'html.parser')
+
+    # Remove unwanted tags entirely
+    for tag in soup(['script', 'style', 'noscript', 'img', 'source', 'picture', 'svg']):
+        tag.decompose()
+
+    # Remove image srcset and media nonsense in attributes
+    for tag in soup.find_all(True):
+        for attr in ["src", "srcset", "data-src", "data-srcset", "alt", "title"]:
+            if attr in tag.attrs:
+                del tag.attrs[attr]
+
+    # Extract visible paragraph-level content
+    paragraphs = soup.find_all(['p', 'h1', 'h2', 'h3', 'li'])
+    raw_text = ' '.join(p.get_text(separator=' ', strip=True) for p in paragraphs)
+
+    # Remove URLs or scene7 junk using regex
+    clean_text = sub(URL_REGEX, "", raw_text)
+    clean_text = sub(SCENE7_JUNK_REGEX, "", clean_text)
+
+    # Reminder: PAGE_LEN_MAX (default 5000) truncates page content earlier in crawl();
+    # if changing size logic, update both places or consolidate here.
+    # Normalize to ascii-only and strip quotes
+    clean_text = clean_text.replace('"', '').replace("'", "")
+    clean_text = clean_text.encode("ascii", errors="ignore").decode()
+
+    return clean_text.strip()
+
+
+def _extract_title_from_html(html: str) -> str:
+    """
+    Extracts the title from the given HTML content using BeautifulSoup.
+
+    This method encapsulates the HTML parsing logic, allowing easier reuse or replacement
+    with a different parser if needed in the future.
+
+    Args:
+        html (str): Raw HTML content.
+
+    Returns:
+        str: The cleaned title string if found, otherwise an empty string.
+    """
+    soup = BeautifulSoup(html, 'html.parser')
+    return soup.title.string.strip() if soup.title and soup.title.string else ""
+
+
+def random_id(prefix="", length=6):
+    """
+    Generates a random alphanumeric identifier with an optional prefix.
+
+    The identifier consists of lowercase letters and digits, with a default total
+    length of 6 characters (excluding the prefix).
+
+    Args:
+        prefix (str): Optional string to prepend to the random ID.
+        length (int): Length of the random portion of the ID (default is 6).
+
+    Returns:
+        str: A string representing the generated ID.
+    """
+    return prefix + ''.join(choices(ascii_lowercase + digits, k=length))
+
+
+def get_agent_network_hocon(agents, agent_network_name):
```

Comment:
Type hints go a long way in setting expectations and debugging. They are a good habit to get into.
Your comments help a lot, tho.

Reply:
Also: It's weird that this method is not a part of your class.